haskell/private/cc_wrapper.bzl

```diff
@@ -41,6 +41,22 @@ _cc_wrapper = rule(
             allow_single_file = True,
         ),
         "platform": attr.string(),
+        # Bazel considers the cc toolchain a build dependency of the cc_wrapper,
+        # but we only need the cc toolchain at runtime. Without relying on
+        # environment variables, this is practically the only way we have to get
+        # our hands on the path to the C compiler other than using environment
+        # variables.
+        #
+        # Using environment variables wouldn't work when using hie-bios.
+        #
+        # Specifying the dependency like this, however, interferes with
+        # cross-compilation, since the provided toolchain will target
+        # the execution platform. In this case, we use environment variables
+        # in the cc_wrapper scripts to indicate the path to the cc compiler.
+        # We are not trying to cross-compile when using hie-bios.
+        #
+        # TODO: Consider using execution groups to transition the toolchain
+        # to the target platform.
```

Comment:
I think this may be a bit hard to understand for someone who has less context about this. I gave it a stab to rephrase this to give a bit more context below.

---

The cc_wrapper provides a thin compatibility layer between GHC and the cc-toolchain that Bazel provides. It is a script that GHC will invoke as the C compiler (`pgmc`) and that itself will invoke the C compiler provided by Bazel's cc-toolchain. Therefore cc_wrapper only needs the cc-toolchain at runtime, i.e. when building Haskell binaries. More importantly, in case of cross-compilation, cc_wrapper needs a cc-toolchain that targets the platform that we are building the Haskell binary for.

However, we do not have control over the arguments that GHC passes to the cc_wrapper in all cases. Meaning, we cannot use command line arguments to configure cc_wrapper to use a particular cc-toolchain. Instead, we use environment variables to pass all relevant information about the cc-toolchain to cc_wrapper.

Unfortunately, environment variables are not available in the IDE use-case, because we cannot specify environment variables at the hie-bios interface. To work around this limitation, we configure a fall-back cc-toolchain as a build-time dependency to cc_wrapper.

Note, Bazel will not resolve the desired cc-toolchain at this build-time dependency in case of cross-compilation. The cc_wrapper is a dependency to the hs-toolchain and Bazel will build it to target the build platform (execute) instead of the ultimate target platform (target). Therefore, the fall-back cc-toolchain will not target the cross-platform and should not be used for any build actions.

TODO: Consider using execution groups to transition the toolchain to the target platform.

Reply:
I didn't expected less of your review. :)

Reply:
Adopted.

Reply:
Thank you! Yes, using the note format is a good idea for this.